amd/src/package_search/mutations.js

```diff
@@ -142,4 +146,57 @@ export default class {
     async changeSort(stateManager, sort, order) {
         await this.searchPackages(stateManager, {sort: sort, order: order}, ["all", "favourites", "mine"]);
     }
+
+    /**
+     * Used to re-/load data of given categories.
+     *
+     * @param {StateManager} stateManager
+     * @param {string[]} categories
+     */
+    async load(stateManager, categories) {
+        await this.searchPackages(stateManager, {}, categories);
+    }
+
+    /**
+     * Used to un-/favourite a package.
+     *
+     * @param {StateManager} stateManager
+     * @param {int} packageid
+     * @param {boolean} favourite
+     */
+    async favourite(stateManager, packageid, favourite) {
+        const state = stateManager.state;
+        try {
+            stateManager.setReadOnly(false);
+            state.general.loading = true;
+            const successful = await favouritePackage(packageid, favourite, this.options.contextid);
+            if (!successful) {
+                return;
+            }
+            for (const category of ["all", "mine", "recentlyused"]) {
+                const pkg = state[`${category}Packages`].get(packageid);
+                if (pkg) {
+                    pkg.isfavourite = favourite;
+                }
+            }
+            stateManager.setReadOnly(true);
+            let page = state.favourites.page;
+            if (!favourite) {
+                // Turn back a page in 'favourites' if the unmarked package was the last one on the page.
+                const isFirstPage = page === 0;
+                const isLastPage = page === Math.floor((state.favourites.total - 1) / this.options.limit);
+                const existsOnePackage = state.favourites.count === 1;
+                if (!isFirstPage && isLastPage && existsOnePackage) {
+                    page -= 1;
+                }
+            }
+            await this.changePage(stateManager, 'favourites', page);
+        } catch (exception) {
+            if (!stateManager.readonly) {
+                state.general.loading = false;
+                stateManager.setReadOnly(true);
+            }
```

Comment:
Bei einem Fehler solltest du `state.general.loading` in jedem Fall wieder auf false setzen. Zur Not musst du noch einmal `stateManager.setReadOnly(false);` ausführen.